numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
I think currently none of the numpy types throw any error in nonzero. But I think in principle they could, or maybe a user-defined type might while not needing the PYAPI.

So I think it is more future-friendly to pass on potential errors rather than throwing our own.

Also, this way the code behaves like the ufunc api loops do.